scripts/components/api-changes-validator/api_usage_statements_generators.ts

```diff
@@ -0,0 +1,249 @@
+      | ts.NodeArray<ts.TypeParameterDeclaration>
+      | undefined
+  ) {}
+
+  /**
+   * @inheritDoc
+   */
+  generate = (): UsageStatements => {
+    if (!this.typeParameters || this.typeParameters.length === 0) {
+      return {};
+    }
+    const parametersWithoutDefaults = this.typeParameters.filter(
+      (parameter) => !parameter.default
+    );
+    if (parametersWithoutDefaults.length === 0) {
+      return {};
+    }
+    const statements: Array<string> = [];
+    for (const typeParameter of parametersWithoutDefaults) {
+      if (typeParameter.constraint) {
+        statements.push(typeParameter.constraint.getText());
+      } else {
+        statements.push('string');
+      }
+    }
+    return { usageStatement: `<${statements.join(', ')}>` };
+  };
+}
+
+/**
+ * Generates import statements from import declarations.
+ *
+ * This is a pass through process. Imports present in API.md report are
+ * required for types defined in that report.
+ */
+export class ImportUsageStatementsGenerator
+  implements UsageStatementsGenerator
+{
+  /**
+   * @inheritDoc
+   */
+  constructor(private readonly node: ts.ImportDeclaration) {}
+  generate = (): UsageStatements => {
+    return {
+      importStatement: this.node.getText(),
+    };
+  };
+}
+
+/**
+ * Generates statements for types defined in API.md.
+ *
+ * The generated statements attempt to define old version of type under new alias
+ * and then attempt to assign old alias to new alias. If new alias does not contain
+ * breaks then this procedure should compile successfully.
+ *
+ * Example:
+ * type SampleTypeBaseline = {}
+ * const SampleTypeUsageFunction = (sampleTypeFunctionParameter: SampleTypeBaseline) => {
+ *     const sampleType: SampleType = sampleTypeFunctionParameter;
+ * }
+ *
+ * Note: using a function closure relieves us from mocking values.
+ */
+export class TypeUsageStatementsGenerator implements UsageStatementsGenerator {
+  /**
+   * @inheritDoc
+   */
+  constructor(
+    private readonly typeAliasDeclaration: ts.TypeAliasDeclaration,
+    private readonly packageName: string
+  ) {}
+  generate = (): UsageStatements => {
+    const typeName = this.typeAliasDeclaration.name.getText();
+    const constName = toLowerCamelCase(typeName);
+    const genericTypeParametersDeclaration =
+      new GenericTypeParameterDeclarationUsageStatementsGenerator(
+        this.typeAliasDeclaration.typeParameters
+      ).generate().usageStatement ?? '';
+    const genericTypeParameters =
+      new GenericTypeParameterUsageStatementsGenerator(
+        this.typeAliasDeclaration.typeParameters
+      ).generate().usageStatement ?? '';
+    const baselineTypeName = `${typeName}Baseline`;
+    const functionParameterName = `${constName}FunctionParameter`;
+    // declare type with same content under different name.
+    let usageStatement = `type ${baselineTypeName}${genericTypeParametersDeclaration} = ${this.typeAliasDeclaration.type.getText()}${EOL}`;
```

Comment:
instead of doing all of this string building ourselves, I wonder if we could use the [typescript printer](https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API#creating-and-printing-a-typescript-ast) to construct this from AST nodes?

Reply:
I did an experiment here https://github.com/sobolk/amplify-backend/commit/c31f73b8937c3f08a4c1395cd812ca09576b0128 .

Which led me to conclusion that using `ts.printer` and `ts.factory` for this purpose is going to be overkill.
I had to spin up extra workspace where I was using expected output and debugger to figure out how to construct desired AST. I covered couple of edge cases around type usage snippet generation but gave up at some point given complexity (the experiment is missing function parameter type mandatory template params).
Given expected low complexity of "usage samples" we generate and high complexity of figuring out how to generate them I think we should keep template approach for now until we discover it doesn't scale well.

Since this is internal detail of the tool and usage snippet generation is under "unit-ish" test control it should be straightforward (but dense) effort to migrate later.